RLFD: fix logical error to avoid false positives

simonschdev opened the pull request:
According to the paper, RLFD has no false positives by design. However, there were false positives when running the test for the RLFD detector. This was mostly due to a logical error in the thresholding at the lowest counter level: The counter at the alternative index (cf. Cuckoo hashing) was always checked, even when the currently checked flow was not hashed to that counter. If there was (another) misbehaving flow at that alternative index, RLFD would report that the currently checked flow was misbehaving, leading to false positives. However, there was an easy fix for that.

Moreover, the allowed rate gamma was cut to zero (by converting it to uint32), leading to distortion of results.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/netsec-ethz/lfd/7)
<!-- Reviewable:end -->

haowu-uiuc commented on caida/caida_test.go:
```diff
@@ -44,7 +44,7 @@ var (
     // trace for testing
     trace *TraceData
-    pcapFilename = "../resource/10K-test-pkts"
+    pcapFilename = "../resource/equinix-chicago.dirA.20160406-125912.UTC.anon.pcap"
```
this line should not be changed. (probably you haven't rebase to the newest master branch)

haowu-uiuc commented on clef/clef.go:
```diff
@@ -29,7 +29,7 @@ type ClefDtctr struct {
 }
 
-func NewClefDtctr(linkCap float64, alpha, beta_th, beta, gamma uint32, t_l time.Duration) *ClefDtctr {
+func NewClefDtctr(linkCap float64, alpha, beta_th, beta uint32, gamma, t_l time.Duration) *ClefDtctr {
```
The current CLEF constructor is not correct, and I am trying to fix it after we make RLFD run correctly. Because we should config EARDet according to given beta_l, gamma_l, and gamma_h (see the NewConfigedEardetDtctr in eardet.go, which was added by me today), so the current CLEF constructor is not correct.

I think maybe a better way is to pass one eardet and two RLFDs as the parameters for CLEF constructor. If you have a chance to fix this, that would be nice, otherwise I will take care next weekend.

haowu-uiuc commented:
Thank you for working on this! Please see my comments in this PR.

Thanks
Hao

> On Oct 9, 2017, at 12:01 AM, simonschdev <notifications@github.com> wrote:
> 
> @simonschdev <https://github.com/simonschdev> requested your review on: netsec-ethz/lfd#7 <https://github.com/netsec-ethz/lfd/pull/7> RLFD: fix logical error to avoid false positives.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/netsec-ethz/lfd/pull/7#event-1283560167>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ASUe3HC_ckgwciIHJbaoe14TLnhFLhNWks5sqcS0gaJpZM4PyEPt>.
>

simonschdev commented:
Thank you for your review, Hao!

I fixed the reference to the pcap file back to the one you are using. Did @hosslen already give you the trace she was using for the report? Otherwise, I can hand it to you via Google Drive.

Moreover, I rewrote the CLEF implementation such that it now expects the subdetectors as arguments. Could you look at it and tell me if that is what you wanted?

Best,
Simon

haowu-uiuc commented:
Nice fixes!

haowu-uiuc commented:
Sorry for the review delay. I was busy on the work in company during the weekday. I am trying to do some work today.
